File: public/codigos/cnpj.js
function calculaCnpjDv(cnpj) {
    // Remove máscara
    cnpj = cnpj.replace(/[^A-Za-z0-9]/g, '').toUpperCase();
    
    if (cnpj.length < 12) {
        throw new Error("CNPJ inválido (tamanho incorreto)");
    }
    
    const cnpjBase = cnpj.substring(0, 12);
    const pesosDV = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
    
    let somatorioDV1 = 0;
    let somatorioDV2 = 0;
    
    for (let i = 0; i < 12; i++) {
        const asciiDigito = cnpjBase.charCodeAt(i) - 48;
        somatorioDV1 += asciiDigito * pesosDV[i + 1];
        somatorioDV2 += asciiDigito * pesosDV[i];
    }
    
    const restoDV1 = somatorioDV1 % 11;
    const dv1 = restoDV1 < 2 ? 0 : 11 - restoDV1;
    
    somatorioDV2 += dv1 * pesosDV[12];
    
    const restoDV2 = somatorioDV2 % 11;
    const dv2 = restoDV2 < 2 ? 0 : 11 - restoDV2;
    
    return `${dv1}${dv2}`;
}

function validaCnpj(cnpj) {
    cnpj = cnpj.replace(/[^A-Za-z0-9]/g, '');
    
    if (cnpj.length !== 14) return false;
    
    const cnpjBase = cnpj.substring(0, 12);
    const dvFornecido = cnpj.substring(12, 14);
    
    if (cnpjBase === '000000000000') return false;
    
    try {
        const dvCalculado = calculaCnpjDv(cnpjBase);
        return dvFornecido === dvCalculado;
    } catch (e) {
        return false;
    }
}

// Para NodeJS
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { calculaCnpjDv, validaCnpj };
}

File: public/codigos/CNPJ.java
public class CNPJ {

    public static String calculaCnpjDv(String cnpj) {
        String cnpjBase = cnpj.replaceAll("[^A-Za-z0-9]", "").toUpperCase();
        if (cnpjBase.length() < 12) {
            throw new IllegalArgumentException("CNPJ inválido (tamanho incorreto)");
        }
        
        cnpjBase = cnpjBase.substring(0, 12);
        int[] pesosDV = {6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2};
        
        int somatorioDV1 = 0;
        int somatorioDV2 = 0;
        
        for (int i = 0; i < 12; i++) {
            int asciiDigito = cnpjBase.charAt(i) - 48;
            somatorioDV1 += asciiDigito * pesosDV[i + 1];
            somatorioDV2 += asciiDigito * pesosDV[i];
        }
        
        int restoDV1 = somatorioDV1 % 11;
        int dv1 = restoDV1 < 2 ? 0 : 11 - restoDV1;
        
        somatorioDV2 += dv1 * pesosDV[12];
        
        int restoDV2 = somatorioDV2 % 11;
        int dv2 = restoDV2 < 2 ? 0 : 11 - restoDV2;
        
        return String.valueOf(dv1) + String.valueOf(dv2);
    }

    public static boolean validaCnpj(String cnpj) {
        String limpo = cnpj.replaceAll("[^A-Za-z0-9]", "").toUpperCase();
        if (limpo.length() != 14 || limpo.startsWith("000000000000")) {
            return false;
        }
        
        String cnpjBase = limpo.substring(0, 12);
        String dvFornecido = limpo.substring(12, 14);
        
        try {
            String dvCalculado = calculaCnpjDv(cnpjBase);
            return dvFornecido.equals(dvCalculado);
        } catch (Exception e) {
            return false;
        }
    }
}
